Add AsrSignal support

YevgenL opened the pull request.

deruelle commented on restcomm/restcomm.docs/sources-asciidoc/src/main/asciidoc/configuration/asr/asr.adoc:
```diff
@@ -0,0 +1,13 @@
+= Automatic Speech Recognition Integration
+
+In order to enable ASR functionality an appropriate configuration has to be added. There are list of available drivers supported by mediaserver and a default one, which can be used by Restcomm. These are environment variables you need to set in restcomm.conf file
+
```
why don't we put those options in https://github.com/RestComm/Restcomm-Connect/blob/master/restcomm/configuration/config-scripts/as7-config-scripts/restcomm/mediaserver.conf instead of restcomm.conf ?

hamsterksu replied:
@deruelle it's configuration for restcomm. this list will be displayed on the account page

gvagenas commented on restcomm/restcomm.interpreter/src/test/java/org/restcomm/connect/interpreter/rcml/GatherSpeechTest.java:
```diff
@@ -0,0 +1,392 @@
+
+            @Override
+            public Actor create() throws Exception {
+                return new VoiceInterpreter(configuration,
```
@YevgenL The VoiceInterpreter constructor has changed and this needs to be fixed here

hamsterksu commented on restcomm/restcomm.mgcp/pom.xml:
```diff
@@ -73,6 +73,12 @@
 			<scope>provided</scope>
 		</dependency>
 
+	  <dependency>
```
@gvagenas do you mean apache commons? 
we are using OctedString class from this lib (`snmp4j `) on mediaserver too. need to change on both sides

YevgenL commented:
Checked `AsrSignalTest`. The problem is in expected result. I removed dependency to snmp4j lib as was requested and used `Hex` from `apache.commons`. This lib encodes without `:`. I modified expected result.

gvagenas commented:
Great work @YevgenL. You work has been merge to branch `asr` for testing

maria-farooq commented on restcomm/restcomm.mscontrol.mms/src/main/java/org/restcomm/connect/mscontrol/mms/MgcpMediaGroup.java:
```diff
@@ -220,29 +232,28 @@ protected void play(final Object message) {
         if (originator != null)
             this.originator.tell(event, self);
-        ivrInUse = false;
+        if (!response.get().isPartial()) {
```
This is unsafe and can cause NPE

Here is a sample

*MediaServer logs*

```
11:25:22,508 DEBUG [SubMgcpTransactionManager] Closed transaction 147483695 for call agent /10.0.0.4:2727 with code 510
11:25:22,508 DEBUG [MgcpChannel] Outgoing MGCP message to /10.0.0.4:2727:

510 147483695 The transaction could not be executed, because some unspecified protocol error was detected
```

*Corresponding restcomm behaviour*

```
11:25:22,512 INFO  [org.restcomm.connect.mscontrol.mms.MgcpMediaGroup] (RestComm-akka.actor.default-dispatcher-150) ********** Media Group akka://RestComm/user/$h/$l/$a/$a Current State: "active
11:25:22,512 INFO  [org.restcomm.connect.mscontrol.mms.MgcpMediaGroup] (RestComm-akka.actor.default-dispatcher-150) ********** Media Group akka://RestComm/user/$h/$l/$a/$a Processing Message: "org.restcomm.connect.mgcp.IvrEndpointResponse sender : class akka.actor.LocalActorRef
11:25:22,512 INFO  [org.restcomm.connect.interpreter.VoiceInterpreter] (RestComm-akka.actor.default-dispatcher-150)  ********** VoiceInterpreter's akka://RestComm/user/$h/$k Current State: gathering
, Processing Message: org.restcomm.connect.mscontrol.api.messages.MediaGroupResponse
11:25:22,512 ERROR [org.restcomm.connect.commons.faulttolerance.RestcommSupervisorStrategy] (RestComm-akka.actor.default-dispatcher-153) RestcommSupervisorStrategy, actor exception handling. Actor path akka://RestComm/user/$h/$l/$a/$a, exception cause java.lang.NullPointerException, default exception handling strategy Resume
11:25:22,512 DEBUG [org.mobicents.servlet.sip.core.session.SipSessionImpl] (RestComm-akka.actor.default-dispatcher-154) dialog associated with this session to create the new request BYE within that dialog gov.nist.javax.sip.stack.SIPDialog@eebc2806
11:25:22,512 DEBUG [org.mobicents.servlet.sip.core.session.SipSessionImpl] (RestComm-akka.actor.default-dispatcher-154) dialog state Confirmed Dialog for that dialog gov.nist.javax.sip.stack.SIPDialog@eebc2806
11:25:22,512 ERROR [org.restcomm.connect.mscontrol.mms.MgcpMediaGroup] (RestComm-akka.actor.default-dispatcher-150) null: java.lang.NullPointerException
	at org.restcomm.connect.mscontrol.mms.MgcpMediaGroup.notification(MgcpMediaGroup.java:254) [restcomm-connect.mscontrol.mms-8.2.0.136.jar:8.2.0.136]
	at org.restcomm.connect.mscontrol.mms.MgcpMediaGroup.onReceive(MgcpMediaGroup.java:357) [restcomm-connect.mscontrol.mms-8.2.0.136.jar:8.2.0.136]
	at akka.actor.UntypedActor$$anonfun$receive$1.applyOrElse(UntypedActor.scala:159) [akka-actor_2.10-2.1.2.jar:]
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:425) [akka-actor_2.10-2.1.2.jar:]
	at akka.actor.ActorCell.invoke(ActorCell.scala:386) [akka-actor_2.10-2.1.2.jar:]
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:230) [akka-actor_2.10-2.1.2.jar:]
	at akka.dispatch.Mailbox.run(Mailbox.scala:212) [akka-actor_2.10-2.1.2.jar:]
	at akka.dispatch.ForkJoinExecutorConfigurator$MailboxExecutionTask.exec(AbstractDispatcher.scala:506) [akka-actor_2.10-2.1.2.jar:]
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:262) [scala-library-2.10.1.jar:]
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:975) [scala-library-2.10.1.jar:]
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1478) [scala-library-2.10.1.jar:]
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:104) [scala-library-2.10.1.jar:]

```

YevgenL replied:
Changed to `if (response.get() == null || !response.get().isPartial())`
https://github.com/RestComm/Restcomm-Connect/pull/2390/commits/b2d39409adf99b27d3b2d95dd3ea19b90283b953

maria-farooq commented on restcomm/restcomm.docs/sources-asciidoc/src/main/asciidoc/rcml/gather-rcml.adoc:
```diff
@@ -1,53 +1,73 @@
+[cols=",",options="header",]
+|=======================================================================
+|Parameter |Description
+|UnstableSpeechResult |Partially recognized speech the caller said.
```
Name of this parameter *UnstableSpeechResult* is confusing. we should make this *PartialSpeechResult*.

maria-farooq commented on restcomm/restcomm.interpreter/src/test/java/org/restcomm/connect/interpreter/GatherSpeechTest.java:
```diff
@@ -0,0 +1,400 @@
+
+                callback = expectMsgClass(HttpRequestDescriptor.class);
+                assertEquals(callback.getUri(), partialCallbackUri);
+                assertEquals(findParam(callback.getParameters(), "UnstableSpeechResult").getValue(), "1");
```
Name of this parameter *UnstableSpeechResult* is confusing. we should make this *PartialSpeechResult*.

maria-farooq commented on restcomm/restcomm.interpreter/src/test/java/org/restcomm/connect/interpreter/GatherSpeechTest.java:
```diff
@@ -0,0 +1,400 @@
+
+                callback = expectMsgClass(HttpRequestDescriptor.class);
+                assertEquals(callback.getUri(), partialCallbackUri);
+                assertEquals(findParam(callback.getParameters(), "UnstableSpeechResult").getValue(), "12");
```
Name of this parameter *UnstableSpeechResult* is confusing. we should make this *PartialSpeechResult*.

maria-farooq commented on restcomm/restcomm.interpreter/src/test/java/org/restcomm/connect/interpreter/GatherSpeechTest.java:
```diff
@@ -0,0 +1,400 @@
+
+                callback = expectMsgClass(HttpRequestDescriptor.class);
+                assertEquals(callback.getUri(), partialCallbackUri);
+                assertEquals(findParam(callback.getParameters(), "UnstableSpeechResult").getValue(), "1");
```
Name of this parameter *UnstableSpeechResult* is confusing. we should make this *PartialSpeechResult*.

maria-farooq commented on restcomm/restcomm.interpreter/src/test/java/org/restcomm/connect/interpreter/GatherSpeechTest.java:
```diff
@@ -0,0 +1,400 @@
+
+                callback = expectMsgClass(HttpRequestDescriptor.class);
+                assertEquals(callback.getUri(), partialCallbackUri);
+                assertEquals(findParam(callback.getParameters(), "UnstableSpeechResult").getValue(), "12");
```
Name of this parameter *UnstableSpeechResult* is confusing. we should make this *PartialSpeechResult*.

maria-farooq commented on restcomm/restcomm.interpreter/src/main/java/org/restcomm/connect/interpreter/BaseVoiceInterpreter.java:
```diff
@@ -1637,6 +1725,25 @@ public void execute(final Object message) throws Exception {
+            final MediaGroupResponse<CollectedResult> asrResponse = (MediaGroupResponse<CollectedResult>) message;
+
+            final List<NameValuePair> parameters = parameters();
+            parameters.add(new BasicNameValuePair("UnstableSpeechResult", asrResponse.get().getResult()));
```
Name of this parameter *UnstableSpeechResult* is confusing. we should make this *PartialSpeechResult*.
Also we can declare it as private static final string.
